_posts/en/2017-01-11-android-listeners.md

```diff
@@ -0,0 +1,173 @@
+        manager.setListener(this);
+        manager.get("http://plop.fr");
+    }
+
+    private void callWebService() {
+        RequestManager manager = new RequestManager();
+        manager.setListener(this);
+        manager.get("http://plop.fr");
+    }
+
+    @Override
+    public void onRequestSuccess() {
+
+    }
+
+    @Override
+    public void onRequestFailure() {
+
+    }
+}
+```
+
+If now, I call the callWebService method, considering the dummy code that we produced, the result will be that we reach the onRequestSuccess method.
+
+**Bu why do that?**
+
+We have a light dependency between our different objects. Indeed, RequestManager has no idea idea of the type of object of its listener, all it cares about is to know that the delegate contains the two methods of the interface so that it is able to call them both. So, it's good, I was able to call my Webservice, I have my two callbacks where I can process what needs to be, everything seems good. Yes everything is good, it works fine and it will cover a lot of cases.
+
+**Raised issue**
+
+In the previous article, I told you about the case where you need to do many calls in the same class. Then you will tell me:
+- There must be something equivalent from closures on Android, like on iOS.
+- Yes, there is one.
+- So, give it to us then, instead of making us wait for it.
+- Listeners
+- What? What is he talking about? We just talked about it !!!
+- Let me explain.
+
+**Solution**
+
+Actually, thanks to Java, you can instantiate an interface if you redefine the methods when you instanciate it. A small sample of code in order to show you?
+
+```Java
+public class MainActivity extends AppCompatActivity {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+        RequestManager manager = new RequestManager();
+        RequesterListener listener = new RequesterListener() {
+
+            @Override
+            public void onRequestSuccess() {
+
+            }
+
+            @Override
+            public void onRequestFailure() {
+
+            }
+        };
+
+        manager.setListener(listener);
+        manager.get("http://plop.fr");
+    }
+}
+```
+### Conclusion
+
+If now, I call the callWebService method, considering the dummy code that we produced, the result will be that we reach the onRequestSuccess method. So, it's pretty cool no? It covers 100% of cases and easily without having to change a lot, you can become listener by inheriting an interface or just define an instance of this one. Personnaly, when I discovered this, I told myself, how can it be? It is just awesome! I look for something equivalent for iOS, telling myself of course they thought about that... Big fail, I found nothing... Then by searching on internet, for a really specific need, I found a pretty elegant solution. But that, it will be covered in another article :) See you space cowboys!
```

Comment:
it's pretty cool, right?
you can set up a listener... defining
I looked for
on the internet
But that will be covered...